fabtests/pytest/common.py

```diff
@@ -300,20 +300,30 @@ def prepare_base_command(self, command_type, executable,
 
         return command
 
-    def _run_client_command(self, server_process):
+    def _run_client_command(self, server_process, client_command, client_output_file=None,
+                            run_client_asynchronously=False):
 
         if server_process.poll():
+            if client_output_file:
+                with open(client_output_file, "w") as f:
+                    f.write("")
```

Comment:
Similar logic can be simplified/more robust if we change function signature to
```python
from typing import IO, Optional # Probably not supported in 3.6 - just to show type.
...
def _run_client_command(self, ..., client_output_file: Optional[IO],..):
    ...
    process = Popen(..., stdout=client_output_file, stderr=subprocess.STDOUT, ....)
    ...
```
In this case the caller needs to open and close the file handles, e.g.
```python
        for i in range(self.numclient):
            client_outfile_handle_list[i] = NamedTemporaryFile(prefix="fabtests_client_{}.out.".format(i))
            ...
            try:
                # Start client
                # Retry on SSH connection error until server timeout
                client_process_list[i] = retry(
                    retry_on_exception=is_ssh_connection_error,
                    stop_max_delay=self._timeout * 1000,  # Convert to milliseconds
                    wait_fixed=CLIENT_RETRY_INTERVAL_MS,
                )(self._run_client_command)(server_process, client_command,
                                            client_outfile_handle_list[i],
                                            self._run_client_asynchronously)
    ....
    for f in client_outfile_handle_list:
        f.close()
```

Reply:
Since  this requires certain python version, can we make it a separate change in the future?

Reply:
Just to be clear, we don't need to import typing which requires a higher version - it was provided FYI. We can still implement the logic though.

Reply:
I am still not sure what's the benefits of using this handle?

Reply:
Passing the file handle to the client process means output is captured in the file automatically.

Therefore, we do not need to handle edge cases when the process is interrupted(or not) and write output manually to the output file, e.g.
```python
if client_output_file:
    # Currently this has to be added manually case by case.
    # Instead, pass the output file to subprocess and let it write to the file directly.
    with open(...):
        ...
```

Reply:
`client_outfile_handle_list[i] = NamedTemporaryFile(prefix="fabtests_client_{}.out.".format(i)` this is filename, not handle?

Reply:
you mean `open(NamedTemporaryFile(prefix="fabtests_client_{}.out.".format(i))`

Reply:
One blocker is if I use the client_outfile_handle as stdout, there will be no output in `process.communicate`, and the following code that use `output` will fail as it is None.

Reply:
> One blocker is if I use the client_outfile_handle as stdout, there will be no output in `process.communicate`, and the following code that use `output` will fail as it is None.

I see. In that case it is troublesome to use the file handle. I'm ok with passing file name then.